--- a/net/minecraft/world/chunk/BlockStatePaletteLinear.java
+++ b/net/minecraft/world/chunk/BlockStatePaletteLinear.java
@@ -27,7 +27,7 @@
         this.deserializer = p_i48962_4_;
     }
 
-    public int idFor(T state)
+    public int getId(T state)
     {
         for (int i = 0; i < this.arraySize; ++i)
         {
@@ -37,6 +37,16 @@
             }
         }
 
+        return -1;
+    }
+
+    public int idFor(T state)
+    {
+        int i = this.getId(state);
+
+        if (i != -1)
+            return i;
+
         int j = this.arraySize;
 
         if (j < this.states.length)
