--- a/net/minecraft/world/chunk/BlockStatePaletteRegistry.java
+++ b/net/minecraft/world/chunk/BlockStatePaletteRegistry.java
@@ -17,9 +17,14 @@
         this.defaultState = p_i48965_2_;
     }
 
+    public int getId(T state)
+    {
+        return this.registry.get(state);
+    }
+
     public int idFor(T state)
     {
-        int i = this.registry.get(state);
+        int i = this.getId(state);
         return i == -1 ? 0 : i;
     }
 
