--- a/net/minecraft/entity/Entity.java
+++ b/net/minecraft/entity/Entity.java
@@ -720,17 +720,21 @@
 
             if (x != 0.0D || y != 0.0D || z != 0.0D)
             {
-                ReuseableStream<VoxelShape> reuseablestream = new ReuseableStream<VoxelShape>(this.world.getCollisionBoxes(this, this.getBoundingBox(), x, y, z));
+                final ocd.mcoptimizations.shapecast.IShapecaster shapecaster = this.world.createShapecaster(this, this.getBoundingBox(), x, y, z);
 
                 if (y != 0.0D)
                 {
-                    y = VoxelShapes.func_212437_a(EnumFacing.Axis.Y, this.getBoundingBox(), reuseablestream.func_212761_a(), y);
-                    this.setBoundingBox(this.getBoundingBox().offset(0.0D, y, 0.0D));
+                    y = shapecaster.shapecast(this.getBoundingBox(), EnumFacing.Axis.Y, y);
+
+                    if (y != 0.0D)
+                    {
+                        this.setBoundingBox(this.getBoundingBox().offset(0.0D, y, 0.0D));
+                    }
                 }
 
                 if (x != 0.0D)
                 {
-                    x = VoxelShapes.func_212437_a(EnumFacing.Axis.X, this.getBoundingBox(), reuseablestream.func_212761_a(), x);
+                    x = shapecaster.shapecast(this.getBoundingBox(), EnumFacing.Axis.X, x);
 
                     if (x != 0.0D)
                     {
@@ -740,7 +744,7 @@
 
                 if (z != 0.0D)
                 {
-                    z = VoxelShapes.func_212437_a(EnumFacing.Axis.Z, this.getBoundingBox(), reuseablestream.func_212761_a(), z);
+                    z = shapecaster.shapecast(this.getBoundingBox(), EnumFacing.Axis.Z, z);
 
                     if (z != 0.0D)
                     {
@@ -764,24 +768,24 @@
 
                 if (d2 != 0.0D || y != 0.0D || d4 != 0.0D)
                 {
-                    ReuseableStream<VoxelShape> reuseablestream1 = new ReuseableStream<VoxelShape>(this.world.getCollisionBoxes(this, this.getBoundingBox(), d2, y, d4));
+                    final ocd.mcoptimizations.shapecast.IShapecaster shapecaster = this.world.createShapecaster(this, this.getBoundingBox().expand(d2, y, d4));;
                     AxisAlignedBB axisalignedbb1 = this.getBoundingBox();
                     AxisAlignedBB axisalignedbb2 = axisalignedbb1.expand(d2, 0.0D, d4);
-                    double d9 = VoxelShapes.func_212437_a(EnumFacing.Axis.Y, axisalignedbb2, reuseablestream1.func_212761_a(), y);
+                    double d9 = shapecaster.shapecast(axisalignedbb2, EnumFacing.Axis.Y, y);
 
                     if (d9 != 0.0D)
                     {
                         axisalignedbb1 = axisalignedbb1.offset(0.0D, d9, 0.0D);
                     }
 
-                    double d10 = VoxelShapes.func_212437_a(EnumFacing.Axis.X, axisalignedbb1, reuseablestream1.func_212761_a(), d2);
+                    double d10 = shapecaster.shapecast(axisalignedbb1, EnumFacing.Axis.X, d2);
 
                     if (d10 != 0.0D)
                     {
                         axisalignedbb1 = axisalignedbb1.offset(d10, 0.0D, 0.0D);
                     }
 
-                    double d11 = VoxelShapes.func_212437_a(EnumFacing.Axis.Z, axisalignedbb1, reuseablestream1.func_212761_a(), d4);
+                    double d11 = shapecaster.shapecast(axisalignedbb1, EnumFacing.Axis.Z, d4);
 
                     if (d11 != 0.0D)
                     {
@@ -789,21 +793,21 @@
                     }
 
                     AxisAlignedBB axisalignedbb3 = this.getBoundingBox();
-                    double d12 = VoxelShapes.func_212437_a(EnumFacing.Axis.Y, axisalignedbb3, reuseablestream1.func_212761_a(), y);
+                    double d12 = shapecaster.shapecast(axisalignedbb3, EnumFacing.Axis.Y, y);
 
                     if (d12 != 0.0D)
                     {
                         axisalignedbb3 = axisalignedbb3.offset(0.0D, d12, 0.0D);
                     }
 
-                    double d13 = VoxelShapes.func_212437_a(EnumFacing.Axis.X, axisalignedbb3, reuseablestream1.func_212761_a(), d2);
+                    double d13 = shapecaster.shapecast(axisalignedbb3, EnumFacing.Axis.X, d2);
 
                     if (d13 != 0.0D)
                     {
                         axisalignedbb3 = axisalignedbb3.offset(d13, 0.0D, 0.0D);
                     }
 
-                    double d14 = VoxelShapes.func_212437_a(EnumFacing.Axis.Z, axisalignedbb3, reuseablestream1.func_212761_a(), d4);
+                    double d14 = shapecaster.shapecast(axisalignedbb3, EnumFacing.Axis.Z, d4);
 
                     if (d14 != 0.0D)
                     {
@@ -828,7 +832,7 @@
                         this.setBoundingBox(axisalignedbb3);
                     }
 
-                    y = VoxelShapes.func_212437_a(EnumFacing.Axis.Y, this.getBoundingBox(), reuseablestream1.func_212761_a(), y);
+                    y = shapecaster.shapecast(this.getBoundingBox(), EnumFacing.Axis.Y, y);
 
                     if (y != 0.0D)
                     {
